### PR TITLE
Adding layers for Sluice Networks

### PR DESCRIPTION
**Pull request by galenxing**
AlphaShare and BetaShare layers provide alpha parameters and beta parameters from sluice networks respectively. 
SluiceLoss is the loss function for sluice networks. 

Also included a LayerSplitter method similar to that explained in #780 but can be used for other layers as well. LayerSplitter assumes the out_tensor is an array of tensors and given the int output_num, returns out_tensor[output_num, :]. Curious on your thoughts on this implementation. 

Ref: https://arxiv.org/abs/1705.08142

### CODE REVIEW

File: deepchem/models/tensorgraph/layers.py [hunk 1]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
     R = tf.reduce_sum(tf.multiply(D, D), 3)
     R = tf.sqrt(R)
     return R
+
+
+def AlphaShare(**kwargs):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Could you make `in_layers` an explicit keyword argument and add to docstring?

File: deepchem/models/tensorgraph/layers.py [hunk 2]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+  num_outputs = len(kwargs['in_layers'])
+  for num_layer in range(0, num_outputs):
+    output_layers.append(
+        LayerSplitter(in_layers=alpha_share, output_num=num_layer))
+  return output_layers
+
+
+class AlphaShareLayer(Layer):
+  """
+    Part of a sluice network. Adds alpha parameters to control
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: The indentation here looks off. You might have an old version of `yapf`. Try `pip install --upgrade yapf` and rerunning.

File: deepchem/models/tensorgraph/layers.py [hunk 3]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+  output_layers = []
+  alpha_share = AlphaShareLayer(**kwargs)
+  num_outputs = len(kwargs['in_layers'])
+  for num_layer in range(0, num_outputs):
+    output_layers.append(
+        LayerSplitter(in_layers=alpha_share, output_num=num_layer))
+  return output_layers
+
+
+class AlphaShareLayer(Layer):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Could you add a unit test for `AlphaShareLayer`?

File: deepchem/models/tensorgraph/layers.py [hunk 4]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+    self.alphas = None
+    return num_outputs, out_tensor, alphas
+
+  def set_tensors(self, tensor):
+    self.num_outputs, self.out_tensor, self.alphas = tensor
+
+
+class LayerSplitter(Layer):
+  """
+    Returns the nth output of a layer
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Same indentation issue here.

File: deepchem/models/tensorgraph/layers.py [hunk 5]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+    Assumes out_tensor has shape [x, :] where x is the total number of intended output tensors
+    """
+
+  def __init__(self, output_num, **kwargs):
+    """
+        Parameters
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Same indentation issue.

File: deepchem/models/tensorgraph/layers.py [hunk 6]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    inputs = self._get_input_tensors(in_layers)[0]
+    self.out_tensor = inputs[self.output_num, :]
+    return self.out_tensor
+
+
+class SluiceLoss(Layer):
+  """
+    Calculates the loss in a Sluice Network
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Same indentation issue.

File: deepchem/models/tensorgraph/layers.py [hunk 7]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+class AlphaShareLayer(Layer):
+  """
+    Part of a sluice network. Adds alpha parameters to control
+    sharing between the main and auxillary tasks
+    """
+
+  def __init__(self, **kwargs):
+    super(AlphaShareLayer, self).__init__(**kwargs)
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Could you add a docstring here?

File: deepchem/models/tensorgraph/layers.py [hunk 8]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+    self.output_num = output_num
+    super(LayerSplitter, self).__init__(**kwargs)
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    inputs = self._get_input_tensors(in_layers)[0]
+    self.out_tensor = inputs[self.output_num, :]
+    return self.out_tensor
+
+
+class SluiceLoss(Layer):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Could you add a unit test for `SluiceLoss`?

File: deepchem/models/tensorgraph/layers.py [hunk 9]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+      subspaces.append(input_tensor[:, subspace_size:])
+      product = tf.matmul(tf.transpose(subspaces[0]), subspaces[1])
+      # calculate squared Frobenius norm
+      temp.append(tf.reduce_sum(tf.pow(product, 2)))
+    out_tensor = tf.reduce_sum(temp)
+    self.out_tensor = out_tensor
+    return out_tensor
+
+
+class BetaShare(Layer):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Could you add a unit test for `BetaShare`?

File: deepchem/models/tensorgraph/layers.py [hunk 10]
```diff
@@ -2614,3 +2614,166 @@ def distance_matrix(self, D):
...
+    self.num_outputs = None
+    self.out_tensor = None
+    self.alphas = None
+    return num_outputs, out_tensor, alphas
+
+  def set_tensors(self, tensor):
+    self.num_outputs, self.out_tensor, self.alphas = tensor
+
+
+class LayerSplitter(Layer):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Could you add a unit test for `LayerSplitter`?

**Comment by rbharath** [comment]
@galenxing Almost there! Just need to resolve the merge conflicts and add a simple example of constructing a sluice network.

**Comment by rbharath** [comment]
@galenxing Now that #813 is merged, you should have fewer yapf problems. Could you merge in master and try again?

**Comment by galenxing** [comment]
@rbharath yes absolutely!

**Comment by galenxing** [comment]
I was using an older version of tensor_graph.py when originally testing the sluice network on tox21, but after updating tensorgraph, the sluice network is practically not learning anything. I'm still looking into why, but it nonetheless provides an example for creating sluice networks.

**Comment by rbharath** [comment]
@galenxing Looks great! Thanks for the contribution and patience during the review process.

Will look forward to seeing a future PR with a learning sluice model :)

LGTM
